lib/concurrent/delay.rb

```diff
@@ -145,16 +147,15 @@ def wait(timeout = nil)
     # @yield the delayed operation to perform
     # @return [true, false] if success
     def reconfigure(&block)
-      mutex.lock
-      raise ArgumentError.new('no block given') unless block_given?
-      unless @computing
-        @task = block
-        true
-      else
-        false
+      mutex.synchronize do
+        raise ArgumentError.new('no block given') unless block_given?
+        unless @computing
+          @task = block
+          true
+        else
+          false
+        end
```

Comment:
It is using the `#synchronize` method from `Synchronization::Object`, it's just doing so through a level of indirection. See [line 64](https://github.com/ruby-concurrency/concurrent-ruby/blob/ivar-promise/lib/concurrent/delay.rb#L64). `Delay` includes `Obligation` which also includes `Dereferenceable`, the latter of which creates its own mutex within the `#init_mutex` function. I've added an argument to both `#init_obligation` and `#init_mutex` that allows the mutex to be injected. `Delay` passes `self` into those initialization functions, thus allowing the delay object to be its own lock. The `#mutex` method returns `self`. The `mutex.synchronize` call on line 150 is actually `self.synchronize`.